tests/validator/validator.cpp

```diff
@@ -2841,3 +2932,534 @@ TEST(Validator, duplicateIdAll)
+    EXPECT_EQ(errorMessage1, validator->issue(0)->description());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->errorCount());
+    EXPECT_EQ(errorMessage2, validator->issue(0)->description());
+}
+
+TEST(Validator, importInvalidComponentReference)
+{
+    const std::string errorMessage =
+        "Imported component 'imported_component' refers to component 'missing_component' which does not appear in 'component_that_is_invalid.cellml'.";
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/importing_a_component_that_is_not_there.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->issueCount());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->errorCount());
+    EXPECT_EQ(errorMessage, validator->issue(0)->description());
+}
+
+TEST(Validator, unitsErrorInImportedComponent)
+{
+    const std::string errorMessage =
+        "Imported component 'component1' is not valid because:\n"
+        "  -> Component 'component1' importing 'component2' from 'model2.cellml' imports:\n"
+        "  -> Component 'component2' importing 'component3' from 'model3.cellml' which has an error:\n"
+        "   - Variable 'time' in component 'component3' has a units reference 'seconds' which is neither standard nor defined in the parent model.";
+
+    auto model1 = libcellml::Model::create("model1");
+    auto model2 = libcellml::Model::create("model2");
+    auto model3 = libcellml::Model::create("model3");
+
+    auto variable1 = libcellml::Variable::create("time");
+
+    auto units1 = libcellml::Units::create("seconds");
+
+    auto component1 = libcellml::Component::create("component1");
+    auto component2 = libcellml::Component::create("component2");
+    auto component3 = libcellml::Component::create("component3");
+
+    auto imp1 = libcellml::ImportSource::create();
+    auto imp2 = libcellml::ImportSource::create();
+
+    imp1->setModel(model2);
+    imp1->setUrl("model2.cellml");
+    imp2->setModel(model3);
+    imp2->setUrl("model3.cellml");
+
+    model1->addComponent(component1);
+    model2->addComponent(component2);
+    model3->addComponent(component3);
+
+    component3->addVariable(variable1);
+
+    variable1->setUnits(units1);
+
+    component1->setSourceComponent(imp1, "component2");
+    component2->setSourceComponent(imp2, "component3");
+
+    auto validator = libcellml::Validator::create();
+
+    validator->validateModel(model1);
+    EXPECT_EQ(size_t(1), validator->errorCount());
+    EXPECT_EQ(errorMessage, validator->error(0)->description());
+}
+
+TEST(Validator, cImportThatIllustratesBadPractice)
```

Comment:
"cImport..."?

Reply:
It's not clear what this test actually does. Clearly, it's related to bad practice(s?), but which one(s?) are we talking about? Same story for the next two tests (i.e. `zImportThatIllustratesBadPractice` and `zOffsetImportThatIllustratesBadPractice`).

Reply:
it is trying to say in a concise way how one might consider the type of import that is happening.  It is quite an edge case test so you really to have to consider the code in the test to understand exactly what is going on.